src/CommunityToolkit.Maui/PopupService.cs

```diff
@@ -0,0 +1,127 @@
+	/// </summary>
+	/// <param name="serviceProvider">The <see cref="IServiceProvider"/> implementation.</param>
+	public PopupService(IServiceProvider serviceProvider)
+	{
+		this.serviceProvider = serviceProvider;
+	}
+
+	internal static void AddTransientPopup<TPopupView, TPopupViewModel>(IServiceCollection services)
+		where TPopupView : Popup
+		where TPopupViewModel : INotifyPropertyChanged
+	{
+		viewModelToViewMappings.Add(typeof(TPopupViewModel), typeof(TPopupView));
+
+		services.AddTransient(typeof(TPopupView));
+		services.AddTransient(typeof(TPopupViewModel));
+	}
+
+	/// <inheritdoc cref="IPopupService.ShowPopup{TViewModel}()"/>
+	public void ShowPopup<TViewModel>() where TViewModel : INotifyPropertyChanged =>
+		ShowPopup(GetViewModel<TViewModel>());
+
+	/// <inheritdoc cref="IPopupService.ShowPopup{TViewModel}(TViewModel)"/>
+	public void ShowPopup<TViewModel>(TViewModel viewModel) where TViewModel : INotifyPropertyChanged
+	{
+		ArgumentNullException.ThrowIfNull(viewModel);
+
+		var popup = GetPopup(viewModel);
+
+		GetMainPage().ShowPopup(popup);
+	}
+
+	/// <inheritdoc cref="IPopupService.ShowPopup{TViewModel}(IDictionary{string, object})"/>
+	public void ShowPopup<TViewModel>(IDictionary<string, object> query) where TViewModel : IQueryAttributable
+	{
+		ArgumentNullException.ThrowIfNull(query);
+
+		var viewModel = GetViewModel<TViewModel>();
+		viewModel.ApplyQueryAttributes(query);
+
+		var popup = GetPopup(viewModel);
+
+		GetMainPage().ShowPopup(popup);
+	}
+
+	/// <inheritdoc cref="IPopupService.ShowPopupAsync{TViewModel}()"/>
+	public Task<object?> ShowPopupAsync<TViewModel>() where TViewModel : INotifyPropertyChanged =>
+		ShowPopupAsync(GetViewModel<TViewModel>());
+
+	/// <inheritdoc cref="IPopupService.ShowPopupAsync{TViewModel}(TViewModel)"/>
+	public Task<object?> ShowPopupAsync<TViewModel>(TViewModel viewModel) where TViewModel : INotifyPropertyChanged
+	{
+		ArgumentNullException.ThrowIfNull(viewModel);
+
+		var popup = GetPopup(viewModel);
+
+		return GetMainPage().ShowPopupAsync(popup);
+	}
+
+	/// <inheritdoc cref="IPopupService.ShowPopupAsync{TViewModel}(IDictionary{string, object})"/>
+	public Task<object?> ShowPopupAsync<TViewModel>(IDictionary<string, object> query) where TViewModel : IQueryAttributable
+	{
+		ArgumentNullException.ThrowIfNull(query);
+
+		var viewModel = GetViewModel<TViewModel>();
+		viewModel.ApplyQueryAttributes(query);
+
+		var popup = GetPopup(viewModel);
+
+		return GetMainPage().ShowPopupAsync(popup);
+	}
+
+	static Page GetMainPage()
+	{
+		var currentApplication = Application.Current;
+
+		ArgumentNullException.ThrowIfNull(currentApplication);
+
+		var mainPage = currentApplication.MainPage;
+
+		ArgumentNullException.ThrowIfNull(mainPage);
+
+		return mainPage;
+	}
+
+	Popup GetPopup<TViewModel>(TViewModel viewModel)
+	{
+		var viewModelType = typeof(TViewModel);
+		var popup = this.serviceProvider.GetService(viewModelToViewMappings[viewModelType]) as Popup;
```

Comment:
If my Popup view implementation expects a ViewModel injection, this line will create a second instance of the ViewModel, rather than use the one that is passed in to this function.

Reply:
We don't create a new instance of an viewModel. We create an instance of pop-up, what is desired, since pop-up can't be reused.

Reply:
That is a very good point! I wonder what the cleanest option would be here... we could:
- consider scoped lifetimes but that runs the risk of people not registering as scoped or moves the responsibility on us to scope (which may be good or bad depending on how we look at it)
- check if the `BindingContext` of the Popup already has a value once created and if it is of the expected view model type
- any other ideas?

Reply:
If my Popup expects TViewModel as a constructor parameter, it will be created again. 

CsharpBindingPopup.xaml.cs provided in the samples is an example of a popup that would exhibit this behavior.

Reply:
In my implementation, I rely on the TViewModel type to get the TView instance directly.  Then check the BindingContext as you suggest.

1. If the BindingContext is set, then probably safe to move on.
2. If the BindingContext is not set, create the ViewModel and assign it.

Reply:
Yeah I think this should _mostly_ be ok. There could be the odd scenario where the developer has injected and set something different as the `BindingContext` and perhaps we should throw an exception in that scenario to make it clear that something isn't as expected. 

For now I have opted to use this approach. I will try and update the XML documentation to make it as clear as possible what is happening.

Reply:
There are still extra instances of TViewModel being created in the ShowPopup/ShowPopupAsync() methods.  These should be removed and GetViewModel<TViewModel>() should only be called when BindingContext is null